File: simulation/velocity_trajectory.py
#!/usr/bin/env python3
"""Reusable velocity trajectory loading and evaluation for simulation tools."""

from __future__ import annotations

import json
import math
from pathlib import Path
from typing import Any

DEFAULT_TRAJECTORY_DICT: dict[str, Any] = {
    "name": "default_takeoff_hover_land",
    "description": "Simple takeoff, hover, land sequence.",
    "rate_hz": 10,
    "segments": [
        {"type": "hold", "label": "hover", "duration_s": 10, "vx_mps": 0.0, "vy_mps": 0.0, "vz_mps": 0.0}
    ],
}


def resolve_trajectory_path(path: str | Path, base_dir: Path | None = None) -> Path:
    trajectory_path = Path(path)
    if trajectory_path.is_absolute():
        return trajectory_path
    return (base_dir or Path.cwd()) / trajectory_path


def validate_trajectory_file(path: str | Path, base_dir: Path | None = None) -> str:
    trajectory_path = resolve_trajectory_path(path, base_dir)
    if not trajectory_path.exists():
        raise ValueError(f"Trajectory file not found: {trajectory_path}")
    data = json.loads(trajectory_path.read_text(encoding="utf-8"))
    if "segments" not in data:
        raise ValueError(f"Trajectory file missing 'segments' key: {trajectory_path}")
    return str(path)


def load_trajectory(path: str | Path | None, base_dir: Path | None = None) -> dict[str, Any]:
    if not path:
        return DEFAULT_TRAJECTORY_DICT
    trajectory_path = resolve_trajectory_path(path, base_dir)
    data = json.loads(trajectory_path.read_text(encoding="utf-8"))
    if "segments" not in data:
        raise ValueError(f"Trajectory file missing 'segments': {trajectory_path}")
    return data


def lerp(a: float, b: float, u: float) -> float:
    return a + (b - a) * u


def interpolate_keyframes(keyframes: list[dict[str, Any]], t: float) -> tuple[float, float, float]:
    if not keyframes:
        return 0.0, 0.0, 0.0
    ordered = sorted(keyframes, key=lambda item: float(item.get("t", 0.0)))
    if t <= float(ordered[0]["t"]):
        k = ordered[0]
        return float(k.get("vx_mps", 0.0)), float(k.get("vy_mps", 0.0)), float(k.get("vz_mps", 0.0))
    if t >= float(ordered[-1]["t"]):
        k = ordered[-1]
        return float(k.get("vx_mps", 0.0)), float(k.get("vy_mps", 0.0)), float(k.get("vz_mps", 0.0))
    for i in range(len(ordered) - 1):
        a = ordered[i]
        b = ordered[i + 1]
        if float(a["t"]) <= t <= float(b["t"]):
            span = max(float(b["t"]) - float(a["t"]), 1e-6)
            u = (t - float(a["t"])) / span
            return (
                lerp(float(a.get("vx_mps", 0.0)), float(b.get("vx_mps", 0.0)), u),
                lerp(float(a.get("vy_mps", 0.0)), float(b.get("vy_mps", 0.0)), u),
                lerp(float(a.get("vz_mps", 0.0)), float(b.get("vz_mps", 0.0)), u),
            )
    k = ordered[-1]
    return float(k.get("vx_mps", 0.0)), float(k.get("vy_mps", 0.0)), float(k.get("vz_mps", 0.0))


def segment_duration(segment: dict[str, Any]) -> float:
    if segment.get("type") == "velocity_keyframes":
        keyframes = segment.get("keyframes", [])
        if not keyframes:
            return 0.0
        return float(sorted(keyframes, key=lambda item: float(item.get("t", 0.0)))[-1]["t"])
    return float(segment.get("duration_s", 0.0))


def segment_velocity(segment: dict[str, Any], t: float) -> tuple[float, float, float]:
    typ = segment.get("type")
    if typ == "hold":
        return float(segment.get("vx_mps", 0.0)), float(segment.get("vy_mps", 0.0)), float(segment.get("vz_mps", 0.0))
    if typ == "velocity_keyframes":
        return interpolate_keyframes(segment["keyframes"], t)
    if typ == "circle_velocity":
        speed = float(segment.get("speed_mps", 2.0))
        radius = float(segment.get("radius_m", 10.0))
        direction = str(segment.get("direction", "cw")).lower()
        vz = float(segment.get("vz_mps", 0.0))
        theta = speed / max(radius, 1e-6) * t
        sign = -1.0 if direction == "cw" else 1.0
        return speed * math.cos(theta), sign * speed * math.sin(theta), vz
    if typ == "figure8_velocity":
        duration = float(segment["duration_s"])
        speed = float(segment.get("speed_mps", 2.0))
        scale = float(segment.get("scale_m", 10.0))
        vz = float(segment.get("vz_mps", 0.0))
        theta = 2.0 * math.pi * t / max(duration, 1e-6)
        dx = scale * math.cos(theta)
        dy = scale * math.cos(2.0 * theta)
        norm = max(math.hypot(dx, dy), 1e-6)
        return speed * dx / norm, speed * dy / norm, vz
    raise ValueError(f"Unknown trajectory segment type: {typ}")

File: src/behavior/velocity_trajectory.cpp
#include "dedalus/behavior/velocity_trajectory.hpp"

#include <algorithm>
#include <cmath>
#include <stdexcept>
#include <string>

namespace dedalus {
namespace {

constexpr double kPi = 3.14159265358979323846;

double lerp(double a, double b, double u) {
    return a + (b - a) * u;
}

Vec3 interpolate_keyframes(const std::vector<TrajectoryKeyframe>& keyframes, double t_s) {
    if (keyframes.empty()) {
        return Vec3{0.0, 0.0, 0.0};
    }
    if (t_s <= keyframes.front().t_s) {
        return Vec3{keyframes.front().vx_mps, keyframes.front().vy_mps, keyframes.front().vz_mps};
    }
    if (t_s >= keyframes.back().t_s) {
        return Vec3{keyframes.back().vx_mps, keyframes.back().vy_mps, keyframes.back().vz_mps};
    }

    for (std::size_t i = 0; i + 1U < keyframes.size(); ++i) {
        const auto& a = keyframes[i];
        const auto& b = keyframes[i + 1U];
        if (a.t_s <= t_s && t_s <= b.t_s) {
            const double span = std::max(b.t_s - a.t_s, 1e-6);
            const double u = (t_s - a.t_s) / span;
            return Vec3{
                lerp(a.vx_mps, b.vx_mps, u),
                lerp(a.vy_mps, b.vy_mps, u),
                lerp(a.vz_mps, b.vz_mps, u)};
        }
    }

    return Vec3{keyframes.back().vx_mps, keyframes.back().vy_mps, keyframes.back().vz_mps};
}

}  // namespace

VelocityTrajectory::VelocityTrajectory(std::vector<TrajectorySegment> segments)
    : segments_(std::move(segments)) {}

VelocityTrajectory VelocityTrajectory::default_hold(double duration_s) {
    TrajectorySegment hold;
    hold.type = "hold";
    hold.label = "default_hold";
    hold.duration_s = duration_s;
    return VelocityTrajectory{{hold}};
}

bool VelocityTrajectory::empty() const {
    return segments_.empty();
}

std::size_t VelocityTrajectory::size() const {
    return segments_.size();
}

const std::vector<TrajectorySegment>& VelocityTrajectory::segments() const {
    return segments_;
}

const TrajectorySegment& VelocityTrajectory::segment(std::size_t index) const {
    return segments_.at(index);
}

Vec3 VelocityTrajectory::velocity_at(std::size_t segment_index, double segment_elapsed_s) const {
    const auto& segment = segments_.at(segment_index);

    if (segment.type == "hold") {
        return Vec3{segment.vx_mps, segment.vy_mps, segment.vz_mps};
    }
    if (segment.type == "velocity_keyframes") {
        return interpolate_keyframes(segment.keyframes, segment_elapsed_s);
    }
    if (segment.type == "circle_velocity") {
        const double speed = segment.speed_mps > 0.0 ? segment.speed_mps : 2.0;
        const double radius = std::max(segment.radius_m, 1e-6);
        const double sign = segment.direction == "cw" ? -1.0 : 1.0;
        const double theta = speed / radius * segment_elapsed_s;
        return Vec3{
            speed * std::cos(theta),
            sign * speed * std::sin(theta),
            segment.vz_mps};
    }
    if (segment.type == "figure8_velocity") {
        const double duration = std::max(segment.duration_s, 1e-6);
        const double speed = segment.speed_mps > 0.0 ? segment.speed_mps : 2.0;
        const double scale = std::max(segment.scale_m, 1e-6);
        const double theta = 2.0 * kPi * segment_elapsed_s / duration;
        const double dx = scale * std::cos(theta);
        const double dy = scale * std::cos(2.0 * theta);
        const double norm = std::max(std::hypot(dx, dy), 1e-6);
        return Vec3{speed * dx / norm, speed * dy / norm, segment.vz_mps};
    }

    throw std::runtime_error("unknown trajectory segment type: " + segment.type);
}

}  // namespace dedalus
